apollo-router/src/plugins/rhai.rs

```diff
@@ -0,0 +1,507 @@
+        self.get(&ByteString::from(key))
+            .cloned()
+            .unwrap_or_default()
+    }
+}
+
+#[derive(Clone)]
+struct Headers(HeaderMap);
+impl Headers {
+    fn set_header(&mut self, name: String, value: String) -> Result<(), Box<EvalAltResult>> {
+        self.0.append(
+            HeaderName::from_str(&name)
+                .map_err(|err| format!("invalid header name '{name}': {err}"))?,
+            HeaderValue::from_str(&value)
+                .map_err(|err| format!("invalid header value '{value}': {err}"))?,
+        );
+        Ok(())
+    }
+    fn get_header(&mut self, name: String) -> String {
+        self.0
+            .get(&name)
+            .cloned()
+            .map(|h| Some(h.to_str().ok()?.to_string()))
+            .flatten()
+            .unwrap_or_default()
+    }
+}
+
+impl Plugin for Rhai {
+    type Config = Conf;
+
+    fn new(configuration: Self::Config) -> Result<Self, BoxError> {
+        tracing::info!("RHAI {:#?}!", configuration.filename);
+        let engine = Engine::new();
+        let ast = engine.compile_file(configuration.filename)?;
+        Ok(Self { ast })
+    }
+
+    fn router_service(
+        &mut self,
+        service: BoxService<RouterRequest, RouterResponse, BoxError>,
+    ) -> BoxService<RouterRequest, RouterResponse, BoxError> {
+        const FUNCTION_NAME: &str = "router_service";
+        if !self
+            .ast
+            .iter_fn_def()
+            .any(|fn_def| fn_def.name == FUNCTION_NAME)
+        {
+            tracing::debug!("RHAI plugin: no router_service function found");
+            return service;
+        }
+        let this = self.clone();
+        let headers_map = Arc::new(Mutex::new(HeaderMap::new()));
+        let headers_cloned = headers_map.clone();
+        let service = service.map_request(move |request: RouterRequest| {
+            let mut headers = headers_cloned.lock().expect("headers lock poisoned");
+            *headers = request.context.request.headers().clone();
+
+            request
+        });
+
+        service
+            .map_response(move |mut response: RouterResponse| {
+                let extensions = block_on(async { response.context.extensions().await.clone() });
+                let (headers, context) =
+                    match this.run_rhai_script(FUNCTION_NAME, headers_map, extensions) {
+                        Ok(res) => res,
+                        Err(err) => {
+                            return plugin_utils::RouterResponse::builder()
+                                .errors(vec![Error::builder()
+                                    .message(format!("RHAI plugin error: {}", err))
+                                    .build()])
+                                .context(response.context)
+                                .build()
+                                .into();
+                        }
+                    };
+                *(block_on(async { response.context.extensions_mut().await })) = context;
```

Comment:
between this and line 108 it seems to be using `context` and `extensions` interchangeably, this is a bit confusing